File: Top Interview Questions/238. Product of Array Except Self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int n = nums.length;
        int[] result = new int[n];
        result[0] = 1;
        for (int i = 1; i <= n - 1; i++){
            result[i] = result[i-1] * nums[i-1];
        }
        
        for (int i = n - 2; i >= 0; i--){
            result[i] = result[i] * nums[i+1];
            nums[i] *= nums[i+1];
        }
        
        return result;
    }
}

File: Microsoft/Array/238. Product of Array Except Self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        re = [1 for i in range(n)]
        
        # from left
        for i in range(1, n):
            re[i] *= re[i-1] * nums[i-1]
        
        # from right
        R = 1
        for i in range(n-1, -1, -1):
            re[i] *= R
            R *= nums[i]
        
        return re
        